### ISSUE TITLE
Recursive NonNullable Type Inference

### ISSUE THREAD

**Issue by LiamPruditsch** [microsoft/TypeScript#50201]
# Bug Report

### 🔎 Search Terms

`NonNullable<T>`, `NonNullable<NonNullable<T>>`

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/MYGwhgzhAECKCuBTJA5A9gE0QHgCoD5oBvAKGnOmDQDsIAXAJ3mDrQYAoAHeAIxAEtg0AG5gQSAFzR01FPBDg+OAgBpo3PoOjVEADzpSEyROizYZchWCV58hAD7b5IaAF4nCgJTEyFP3QALfggAOlFxRDcRMSRfP3JA4JCdfSiUujjoAF8SHJI9TjY6aCwAMzB5YtBIGCMkWx8-TgZ+UTpI+jA6LQZEMAwaEABPaFwAQQBxKXoW6gBzKIByOsRFzObWrsjShhoDOCRUTBwLZ2sQZTtoR2pndZa2yN6wBkNDk2PzGktFC9sHDwge6bdrQC7zQJSW4AWx4iAYJEyVFojGYrA41GOEAA-FJTlYbARvKR4hREqEdns0ndSQkgqFngxqQpMv56SFwXNAlEAAysij8UrQdiYrAQYn80nkkLcCABEVYzyS7KZPJNXgCISy+XhSTSb5nQn4TxSYRofgYRq05H0aB0aGcKLYFamE4Ggl-Ah2HQAdwOxld7F1iCVtOgguFAEJpZzAt5pZTqMV3PbOMrECAIJFpYzI8k9Mm7Q7lTm+kyU8WwzHEBCAgBqOuq3JAA)

### 💻 Code

```ts
class QueueNode<T> {
    constructor(public value: NonNullable<T>, public next: QueueNode<NonNullable<T>> | null = null) {
        this.value = value
        this.next = next
    }
}

export default class Queue<T> {
    private static readonly TAG: string = 'Queue'
    private front: QueueNode<NonNullable<T>> | null
    private rear: QueueNode<NonNullable<T>> | null
    private length: number

    constructor(nodes?: NonNullable<T>) {
        this.front = null
        this.rear = null
        this.length = 0
        if (nodes) {
            this.push(nodes)
        }
    }

    public push(value: NonNullable<T>): void {
        const tmp = <QueueNode<NonNullable<T>>>new QueueNode(value)
        if (!this.length) this.front = tmp
        else this.rear!.next = tmp
        this.rear = tmp
        this.length++
    }
}
```

### 🙁 Actual behavior

Just discovered NonNullable<T> and tried it out. For some reason, it gives me an error saying `Type 'QueueNode<NonNullable<T>>' is not assignable to type 'QueueNode<NonNullable<NonNullable<T>>>'.` on this line `else this.rear!.next = tmp`
But when I change `NonNullable<T>` to just `T` in the QueueNode constructor, the error is gone and I don't have to infer the type anymore.

### 🙂 Expected behavior
It should not declare the type of NonNullable<T> recursively.

**Comment by RyanCavanaugh** [comment]
This has been fixed in 4.8 with the new definition of `NonNullable<T>`

**Comment by LiamPruditsch** [comment]
Yes, I can confirm. It no longer gives me an error message. I just have to wait until the webpack plugins are updated to finally use it.
